### PR TITLE
Class.Thenable

### PR DESCRIPTION
**Pull request by timwienk**
This is code I made almost a year ago, I finished it (or actually mostly its specs) now we use Mocha, so I could integrate the [Promises/A+ tests][].

New Class mixin. When implemented in a Class, it makes the class "thenable" in the [Promises/A+][] sense of the word, meaning it can be used in Promise style flows by using its `then` method.

The implementation, however, is more than just a "then" method. Any instance of a Class implementing `Class.Thenable` is a `Promises/A+` compliant object (generally referred to as "a Promise") with only one exception: it is possible to reset the Class's value resolution state fully (rejecting pending reactions, and starting empty) to support a Class instance living for longer than just the lifetime of one value resolution.

Example using Request:

```javascript
var request = new Request();
request.send().then(function(response){ console.log(response.text); });
```

Example hooking into a native Promise:

```javascript
var request = new Request();
var promise = Promise.resolve(request);
request.send();
promise.then(function(response){ console.log(response.text); });
```

It's even quite easy to create an actual `Promise` Class if you want, but that seemed a bit redundant to include in MooTools Core. I used the actual way to do this as example in the Docs, though:

```javascript
var Promise = new Class({
    Implements: Class.Thenable,
    initialize: function(executor){
        if (typeof executor !== 'function'){
            throw new TypeError('Promise constructor takes a function argument.');
        }

        try {
            executor(this.resolve.bind(this), this.reject.bind(this));
        } catch (exception){
            this.reject(exception);
        }
    },
    resetThenable: function(){
        throw new TypeError('A promise can only be resolved once.');
    }
});
```

This also solves issue #2643.

Full Travis run:
- https://travis-ci.org/timwienk/mootools-core/builds/84324415

[Promises/A+ tests]: https://www.npmjs.com/package/promises-aplus-tests
[Promises/A+]: https://promisesaplus.com/

### CODE REVIEW

File: Source/Class/Class.Thenable.js [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+var defer;
+if (typeof process !== 'undefined' && typeof process.nextTick === 'function'){
+	defer = process.nextTick;
+} else if (typeof setImmediate !== 'undefined'){
+	defer = setImmediate;
+} else {
+	defer = function(fn){
+		setTimeout(fn, 0);
+	}
```
Comment (anutron) [inline Source/Class/Class.Thenable.js]: `;`

/shrug

**Comment by anutron** [comment]
I guess my only real comment here is if we want this integrated into core. That is to say whether or not this feature needs to be a requirement for Fx, Request, etc. I get it; you can't drop it in on its own (at least, not without something hacky like `Class.Refactor`), so if you want it in Fx/Request you have to put it here, but in some ways this adds bytes to the core library for functionality that maybe isn't primarily useful for most developers.

I'll let others sound off on whether that's the case or not. The code itself looks fine to me. Just a matter of whether we want to add the bytes for this feature.

**Comment by fakedarren** [comment]
I like it. Gzipped the code will be next to nothing, and promises are pretty much required nowadays.
+1 from me. It makes a good blog post!

File: Source/Class/Class.Thenable.js (hunk 1 above)
Comment (timwienk) [inline Source/Class/Class.Thenable.js]: Thanks!

Comment (timwienk) [inline Source/Class/Class.Thenable.js]: Fixed and rebased, surely we don't need a commit for a `;`. :+1:

**Comment by kentaromiura** [comment]
:+1:

**Comment by SergioCrisostomo** [comment]
:+1:

**Comment by arian** [comment]
It's still true that `request.send() === request`?

Isn't it usually that some action, in the case of request `.send()` returns a _new_ promise object, which can be seen as a 'placeholder' until the promise is fulfilled or rejected?

**Comment by timwienk** [comment]
Thanks for the feedback and questions. Quite happy to see practical responses.

@anutron, I agree with the question you put forward, I guess it's quite obvious I agree with @fakedarren's answer, otherwise I wouldn't have written the code, but it's a good thing to bring up. I think this implementation offers something new and very useful by not attempting to add "yet another Promise implementation", but instead offering a way to work with Promise implementations in a compliant manner _and_ still working fine stand-alone.

To answer @arian I'll need a bit more text, because it gets a bit detailed, so brace yourselves (and blame @arian, he asked!):

**[TL;DR]** I'm generally available on IRC if anyone prefers to discuss/ask there in case this text isn't clear or is too long to parse.

Yes, `request.send()` still returns itself. Classes and their instances work as they always have, nothing changed in that respect, and everything works as we expect the Classes to. What this code does is not actually implement Promises itself, it implements a mixin to make any class Promises/A+ compliant (or to actually implement Promises if you want to, see the example in the description of the PR).

There is no rule (or reason for) saying that an arbitrary object cannot act the same way as a Promise, because its goal is to (quite literally) promise that a value will be resolved at some point. You can basically interpret `new Request` as the instantiation of very specific kind of Promise that goes to resolve its value by calling `send`. Another way to look at it is that a `new Request` keeps a Promise internally which you interact with, but the former seems more accurate.

If you do want an actual Promise to work with, you could do `Promise.resolve(request)` (or `Promise.resolve(request.send())` if you feel that's nicer), which works because of [part 2.3.3.3 of the spec][]: the Promise calls `request.then` and uses that to resolve itself (it basically adopts the request's state).

The usual (and even mandatory) thing is that calling `then` does _not_ return itself and returns a new object that gets resolved with the value returned by the argument(s) you pass to `then`, which is indeed what it does.

I do want to point out that referring to Class.Thenable as a Promise implementation is not technically correct, because of the exception I mentioned in the description of this PR, which is related but not the same as what you mention, @arian. As I mentioned, the case is not that `send()` should return a new Promise(-like) object, but it _is_ the case that a Promise can only be used/resolved once. Because Class instances are often used more than once (e.g. multiple `request.send` or `fx.start` calls), this implementation makes the state resettable. Using the "internal Promise" analogue I used earlier: think of it as ditching the old Promise (rejecting it if it's not yet resolved) and taking on a new one. This way nothing is left unresolved, and it's similar to either doing a new `send` after things are done, or canceling the old `send`.

The only thing that does get interesting is using `Class.Thenable` style interaction together with `Chain` style interaction, these are two different ways of chaining actions. I noted this in the docs. (Probably a minor issue anyway, because of defaults and the unlikelihood of people using `chain` if they can use `then`. We could even consider adapting `Chain` to use `Class.Thenable`, but let's not get into that right now.)

[part 2.3.3.3 of the spec]: https://promisesaplus.com/#point-56

File: Source/Class/Class.Thenable.js [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+function reset(thenable){
+	if (thenable.$thenableState !== STATE_PENDING){
+		thenable.$thenableResult = null;
+		thenable.$thenableState = STATE_PENDING;
+	}
+}
+
+function react(thenable){
```
Comment (arian) [inline Source/Class/Class.Thenable.js]: Interesting implementation of _react_ :stuck_out_tongue_winking_eye:

Comment (timwienk) [inline Source/Class/Class.Thenable.js]: :laughing:

For reference, the terminology is based the [ES spec][], except slightly shorter named. This specific function basically implements [TriggerPromiseReactions][], but naming it "react" seemed clear enough to me.

[ES spec]: http://www.ecma-international.org/ecma-262/6.0/#sec-promise-objects
[TriggerPromiseReactions]: http://www.ecma-international.org/ecma-262/6.0/#sec-triggerpromisereactions

**Comment by arian** [comment]
:ship: it

**Comment by anutron** [comment]
I think your description here Tim belongs in the blog post. Yes, it's
fairly technical, but so is our audience (what's left of it).

**Comment by SergioCrisostomo** [comment]
++ for a blog post also from me!

**Pull request merged**